driver/src/main/com/mongodb/reactivestreams/client/internal/GridFSDownloadPublisherImpl.java

```diff
@@ -0,0 +1,214 @@
+                    requestMoreOrComplete();
+                }
+            }
+        }
+
+        @Override
+        public void request(final long n) {
+            synchronized (lock) {
+                requested += n;
+            }
+            requestMoreOrComplete();
+        }
+
+        @Override
+        public void cancel() {
+            terminate();
+        }
+
+        private void requestMoreOrComplete() {
+            synchronized (lock) {
+                if (requested > 0 && !requestedData && !isTerminated && !isCompleted) {
+                    requestedData = true;
+                    if (gridFSFile == null) {
+                        getGridFSFile().subscribe(gridFSFileSubscriber);
+                    } else {
+                        requested--;
+                        int chunkSize = gridFSFile.getChunkSize();
+                        long remaining = gridFSFile.getLength() - sizeRead;
+
+                        if (remaining == 0) {
+                            isCompleted = true;
+                            requestedData = false;
+                            requestMoreOrComplete();
+                        } else {
+                            int byteBufferSize = Math.max(chunkSize, bufferSizeBytes);
+                            byteBufferSize =  Math.min(Long.valueOf(remaining).intValue(), byteBufferSize);
+                            int batchSize = Math.max(byteBufferSize / chunkSize, 1);
+                            ByteBuffer byteBuffer = ByteBuffer.allocate(byteBufferSize);
+                            if (batchSize != currentBatchSize) {
+                                currentBatchSize = batchSize;
+                                gridFSDownloadStream.batchSize(batchSize);
+                            }
+                            gridFSDownloadStream.read(byteBuffer).subscribe(new GridFSDownloadStreamSubscriber(byteBuffer));
+                        }
+                    }
+                } else if (isCompleted && !requestedData && !isTerminated) {
+                    isTerminated = true;
+                    gridFSDownloadStream.close().subscribe(new Subscriber<Success>() {
+                        @Override
+                        public void onSubscribe(final Subscription s) {
+                            s.request(1);
+                        }
+
+                        @Override
+                        public void onNext(final Success success) {
+                        }
+
+                        @Override
+                        public void onError(final Throwable t) {
+                            outerSubscriber.onError(t);
+                        }
+
+                        @Override
+                        public void onComplete() {
+                            outerSubscriber.onComplete();
+                        }
+                    });
+                }
+            }
+        }
+
+        private void terminate() {
+            synchronized (lock) {
```

Comment:
Calling `terminate` should ideally propagate cancellation to the source to stop the producer of the data, if that's possible.

Reply:
Added a check to stop that happening.